File: sem4/jpp/list2/cpp/GF1234577Field.cpp
#include "GF1234577Field.hpp"

#include <stdexcept>
#include <iostream>

GF1234577Field::GF1234577Field(const long long value) : value_(value % getCharacteristic()) {
    if (value_ < 0) {
        value_ += getCharacteristic();
    }
}

long long GF1234577Field::getValue() const {
    return value_;
}

bool GF1234577Field::operator==(const GF1234577Field& other) const {
    return value_ == other.value_;
}

std::strong_ordering GF1234577Field::operator<=>(const GF1234577Field& other) const {
    return value_ <=> other.value_;
}

GF1234577Field GF1234577Field::operator+(const GF1234577Field& other) const {
    return GF1234577Field((value_ + other.value_) % getCharacteristic());
}

GF1234577Field GF1234577Field::operator-(const GF1234577Field& other) const {
    return GF1234577Field((value_ - other.value_) % getCharacteristic());
}

GF1234577Field GF1234577Field::operator*(const GF1234577Field& other) const {
    return GF1234577Field((value_ * other.value_) % getCharacteristic());
}

GF1234577Field GF1234577Field::operator/(const GF1234577Field& other) const {
    if (other.value_ == 0) {
        throw std::invalid_argument("Division by 0!");
    }
    const long long inverse = getModularInverse(other.value_);
    return GF1234577Field((value_ * inverse) % getCharacteristic());
}

GF1234577Field& GF1234577Field::operator=(const GF1234577Field& other) {
    value_ = other.value_;
    return *this;
}

GF1234577Field& GF1234577Field::operator+=(const GF1234577Field& other) {
    value_ = (value_ + other.value_) % getCharacteristic();
    return *this;
}

GF1234577Field& GF1234577Field::operator-=(const GF1234577Field& other) {
    value_ = (value_ - other.value_) % getCharacteristic();
    if (value_ < 0) {
        value_ += getCharacteristic();
    }
    return *this;
}

GF1234577Field& GF1234577Field::operator*=(const GF1234577Field& other) {
    value_ = (value_ * other.value_) % getCharacteristic();
    return *this;
}

GF1234577Field& GF1234577Field::operator/=(const GF1234577Field& other) {
    if (other.value_ == 0) {
        throw std::invalid_argument("Division by 0!");
    }
    const long long inverse = getModularInverse(other.value_);
    value_ = (value_ * inverse) % getCharacteristic();
    return *this;
}

std::ostream& operator<<(std::ostream& outputStream, const GF1234577Field& field) {
    return outputStream << field.value_;
}

GF1234577Field::operator int() const {
    return static_cast<int>(value_);
}

GF1234577Field::operator float() const {
    return static_cast<float>(value_);
}

GF1234577Field::operator double() const {
    return static_cast<double>(value_);
}

long long GF1234577Field::getModularInverse(long long number) const {
    long long modulo = getCharacteristic();
    const long long originalModulo = modulo;
    long long inverse = 0;
    long long tempResult = 1;

    while (number > 1) {
        long long quotient = number / modulo;
        long long temp = modulo;
        modulo = number % modulo;
        number = temp;
        temp = inverse;
        inverse = tempResult - quotient * inverse;
        tempResult = temp;
    }

    if (tempResult < 0)
        tempResult += originalModulo;

    return tempResult;
}

File: sem4/jpp/list2/java/gf1234577field/src/main/java/com/github/kallazz/gf1234577field/GF1234577Field.java
package com.github.kallazz.gf1234577field;

public class GF1234577Field {
    private static final int CHARACTERISTIC = 1234577;
    private int value;

    public GF1234577Field(final int value) {
        this.value = value % CHARACTERISTIC;
        if (this.value < 0) {
            this.value += CHARACTERISTIC;
        }
    }

    public static int getCharacteristic() {
        return CHARACTERISTIC;
    }

    public int getValue() {
        return value;
    }

    private int getModularInverse(int number) {
        int modulo = getCharacteristic();
        final int originalModulo = modulo;
        int inverse = 0;
        int tempResult = 1;

        while (number > 1) {
            int quotient = number / modulo;
            int temp = modulo;
            modulo = number % modulo;
            number = temp;
            temp = inverse;
            inverse = tempResult - quotient * inverse;
            tempResult = temp;
        }

        if (tempResult < 0)
            tempResult += originalModulo;

        return tempResult;
    }

    public boolean equals(GF1234577Field other) {
        return value == other.value;
    }

    public boolean notEquals(GF1234577Field other) {
        return value != other.value;
    }

    public boolean lessThan(GF1234577Field other) {
        return value < other.value;
    }

    public boolean lessThanOrEqual(GF1234577Field other) {
        return value <= other.value;
    }

    public boolean greaterThan(GF1234577Field other) {
        return value > other.value;
    }

    public boolean greaterThanOrEqual(GF1234577Field other) {
        return value >= other.value;
    }

    public GF1234577Field add(GF1234577Field other) {
        return new GF1234577Field((value + other.value) % getCharacteristic());
    }

    public GF1234577Field subtract(GF1234577Field other) {
        return new GF1234577Field((value - other.value) % getCharacteristic());
    }

    public GF1234577Field multiply(GF1234577Field other) {
        return new GF1234577Field((value * other.value) % getCharacteristic());
    }

    public GF1234577Field divide(GF1234577Field other) {
        if (other.value == 0) {
            throw new IllegalArgumentException("Division by 0!");
        }
        final int inverse = getModularInverse(other.value);
        return new GF1234577Field((value * inverse) % getCharacteristic());
    }

    public GF1234577Field assign(GF1234577Field other) {
        value = other.value;
        return this;
    }

    public GF1234577Field assignAdd(GF1234577Field other) {
        value = (value + other.value) % getCharacteristic();
        return this;
    }

    public GF1234577Field assignSubtract(GF1234577Field other) {
        value = (value - other.value) % getCharacteristic();
        if (value < 0) {
            value += getCharacteristic();
        } 
        return this;
    }

    public GF1234577Field assignMultiply(GF1234577Field other) {
        value = (value * other.value) % getCharacteristic();
        return this;
    }

    public GF1234577Field assignDivide(GF1234577Field other) {
        if (other.value == 0) {
            throw new IllegalArgumentException("Divison by 0!");
        }
        final int inverse = getModularInverse(other.value);
        value = (value * inverse) % getCharacteristic();
        return this;
    }

    @Override
    public String toString() {
        return String.valueOf(value);
    }

    public int intValue() {
        return value;
    }

    public float floatValue() {
        return (float) value;
    }

    public double doubleValue() {
        return (double) value;
    }
};
